### PR TITLE
Multiple interpreters and HTML docs testing

### PR DESCRIPTION
**Pull request by flacjacket**
Currently, this is still kind of hacked together and mostly up for review, so any feedback/ways to improve things/how you would like to see this sort of thing work is much appreciated.

This primarily does two things:
- It allows you to specify multiple interpreters by adding multiple `-i` flags. The results for all of the results are uploaded after all of them have run and a summary GitHub comment is made. It currently always creates this summary post, which should probably be changed to do so only when there are multiple interpreters. For an example, see [here](https://github.com/sympy/sympy/pull/1403#issuecomment-6755743).
- Tests the HTML docs building by adding the `-D` flag. This is basically taking what Aaron mentioned in sympy/sympy#1400 and adds it as one of the tests. The main difference is the summary in the GitHub comment is different, it gives the docs building command and the Sphinx version. I'm not sure if there is any other relevant info, but that is what I came up with as possibly useful. There isn't really any error checking to make sure the tester has, for example, sphinx installed. Also, adding the flag adds the docs as an additional test, there isn't a way to do only the docs, which would be good for PR's that only improve docs. For an example, see [here](https://github.com/sympy/sympy/pull/1400#issuecomment-6767169).

### CODE REVIEW

**Comment by certik** [comment]
Nice, the argparse is a much better approach. I am +1 to everything that you did.

Let's test this a little bit. If it works, let's push it in.

Thanks!

**Comment by asmeurer** [comment]
I've already seen the results of this and I like it.  I think the next step forward after this is to have the bot edit the comment to add more tests, rather than posting a new one.  That way we spam much less with it.

Should we wait for sympy/sympy/#1400 to be merged first?  The `-D` flag won't work without it.

**Comment by asmeurer** [comment]
The only downside to this PR is that the bot now requires Python 2.7, but I don't think it's that big of a deal (and probably worth it for the new features). 

Perhaps it should at least be added to the README, though.

**Comment by asmeurer** [comment]
The interactive help is now a little more confusing (though I understand why). Is it possible to make `sympy-bot --help` always also print the help for review?  If not, maybe add some preamble or postamble to make it clear that `sympy-bot review -h` gives a full help.  

Also, is it possible to make `sympy-bot` with no arguments do `sympy-bot -h` as it did before?

**Comment by asmeurer** [comment]
When using argparse,  it's better to use `%(default)s` instead of explicitly typing the default in the help string.

**Comment by asmeurer** [comment]
Why not just make `-D` only build the docs?  If you want to also run the tests, you can do it by adding `-i python` (maybe just a bare `-i` could run the default Python).

**Comment by asmeurer** [comment]
By the way, @certik, if I could get your input on sympy/sympy#1400, that would be great.  The only remaining doc build failure relates to the translations of the tutorial, and I'm not sure how to proceed.

**Comment by flacjacket** [comment]
Alright, so now having `-D` but no `-i` runs just the doc build and you can just have a bare `-i` to also run the tests with the default interpreter.

I set argparse to show the full help when there are too few args. I don't think you can display a subparser help either when it does this and as part of the default help. I also reworded the epilog in the main help to be clear on how to see the other help. Argparse can be installed for older versions of Python, so that won't break it in older versions, which I noted in the README.

Also, I changed how the config file is read in, using ConfigParser instead of exec'ing the file. You will need to change your config file and remove any quotes around strings. In Python 3 it is called configparser, but the bot doesn't run in 3, so I didn't do any handling of the import to do this. Now, you can specify multiple interpreters as default as a comma separated list and all of the command line options can be configured by the config file, tho bools might not be handled quite right.

If you think the docs pull will be in soon, I feel this can wait until then.

**Comment by asmeurer** [comment]
But you've also used things like dictionary comprehensions and set literals (if I remember correctly), which are Python 2.7 only.

**Comment by flacjacket** [comment]
Ah, right. I got rid of the dictionary comprehension. I don't think I used set literals anywhere. The `{}` are empty dicts, not sets, so those are fine.

**Comment by asmeurer** [comment]
You do have a set literal (try running the bot in Python 2.6).  Honestly, though, it isn't that big of a deal if it requires Python 2.7.

**Comment by asmeurer** [comment]
Oh, never mind.  That's just a dictionary with a regular SyntaxError.

**Comment by flacjacket** [comment]
This latest commit adds support for creating various profiles in the config files (e.g. slow tests, 32-bit interpreters, etc.) controlled by the `--profile` option. This will require more changes to the config file, bringing the list up to adding a `[default]` section line and removing quotes around things. I've made the changes clear in the README and decided to include an example config file separate from the README.

I realize I'm pulling configuration in two different direction between this and the API token/config file generation in the other PR, which in retrospect was bad. But I think if I can get one of them merged to master, merging the changes together will make things simpler, rather than messier.

**Comment by asmeurer** [comment]
Is there a way to add a profile for building the docs?

**Comment by asmeurer** [comment]
This is really awesome, by the way. 

@Krastanov you might want to review this if you have the time to see if there are any features that you might want that are missing, as I think this will be a great way to run your never ending bot.

File: sympy-bot [hunk 1]
```diff
@@ -1,229 +1,390 @@
...
+            if message == "too few arguments":
+                self.print_help(sys.stderr)
+            else:
+                self.print_usage(sys.stderr)
+            self.exit(2, '%s: error: %s\n' % (self.prog, message))
+
+    parser = BotArgumentParser(
+            epilog="For the full help on the  review and list commands, run "
+            "'sympy-bot command -h'.",
+            formatter_class=ArgumentDefaultsHelpFormatter)
```
Comment (asmeurer) [inline sympy-bot:41]: Cool, I didn't know you could do this.  argparse is awesome.

**Comment by flacjacket** [comment]
This new commit allows a profile to build the docs. One thing is currently if the config file picks up an interpreter, it will always run the tests, so if you want to be able to just build docs, you can't have a default set of interpreters. Is there some way you'd like to specify for a profile not to run the Python tests? Something like `interpreter = None` or `interpreter =`? Should there be a way to disable Python tests from the command line?

**Comment by asmeurer** [comment]
`interpreter = None` seems reasonable to me.

**Comment by flacjacket** [comment]
Now `interpreter = None` works in the config file and also I guess `-i None` would do the same from the command line.

**Comment by Krastanov** [comment]
When this gets merged it may be interesting to add a rule for it to the travisCI config.

I will pull it now and start using it for the bot that I run.

**Comment by Krastanov** [comment]
The test were ran and at the end I got

```
> View log at: /tmp/sympy-bot-tmpEBf82u/out/log
> Uploading test results for pull number 1293
> Uploaded report for '/usr/bin/python2.7' at: http://reviews.sympy.org/report/agZzeW1weTNyDAsSBFRhc2sYhpgfDA
> Uploaded report for '/usr/bin/python3.2' at: http://reviews.sympy.org/report/agZzeW1weTNyDAsSBFRhc2sYpbcfDA
> Uploaded report for '/usr/local/bin/python2.5' at: http://reviews.sympy.org/report/agZzeW1weTNyDAsSBFRhc2sYjb8fDA
> Uploaded report for 'build docs' at: http://reviews.sympy.org/report/agZzeW1weTNyDAsSBFRhc2sY7p8fDA
Traceback (most recent call last):
  File "./sympy-bot", line 451, in <module>
    main()
  File "./sympy-bot", line 167, in main
    dispatch_reviews(options, urls, username=username, password=password)
  File "./sympy-bot", line 308, in dispatch_reviews
    master_hash, branch_hash, user, config)
  File "./sympy-bot", line 382, in formulate_review
    details = get_sphinx_version()
  File "./sympy-bot", line 445, in get_sphinx_version
    import sphinx
ImportError: No module named sphinx
```

Do we want to catch this exception? It happened when the results were uploaded, not when the test was starting.

**Comment by flacjacket** [comment]
Yeah, that is from the end when it is creating the comment to post on the PR, it tries to import sphinx so it can get the version number (see [here](https://github.com/sympy/sympy/pull/1400#issuecomment-6767169)). Do you have sphinx installed on your bot server? As a side note, testing the html docs won't work until sympy/sympy#1400 is in, that's what sets up the `make html-errors` command.

Nevertheless, we may want to catch that error before trying to build the docs, just to be sure.

**Comment by Krastanov** [comment]
I was just going to report the problem with the `make html-errors` command.

http://reviews.sympy.org/report/agZzeW1weTNyDAsSBFRhc2sY9sYfDA

I have installed sphinx and now it seems that it works

**Comment by flacjacket** [comment]
Now it catches that error and disables the doc building if it can't find sphinx.

**Comment by asmeurer** [comment]
If it will make things easier, we can merge sympy/sympy#1400 (assuming it gets a positive review), and fix the last remaining sphinx error later.

**Comment by Krastanov** [comment]
There seems to be some problem with memory leaks or cash or something like that. When I test using multiple interpreters (instead of running sympy-bot multiple times) the process uses an ungodly amount of RAM.

**Comment by flacjacket** [comment]
> problem with ... cash or something

Mo money mo problems :P

Barring something more insidious, that's probably happening because it holds on to all the reviews while it is reviewing and uploads them all at the end. In that case, it might be better to upload them right away and free up whatever memory we can. I can take a look at it from a memory standpoint today and see what I can come up with.

**Comment by flacjacket** [comment]
@Krastanov, try it now. If the problem was in holding onto all the logs, that should be fixed now. Also, on what order of magnitude is an ungodly amount of memory?

**Comment by Krastanov** [comment]
On 11 July 2012 18:30, Sean Vig
reply@reply.github.com
wrote:

> Stephan, try it now. If the problem was in holding onto all the logs, that should be fixed now. Also, on what order of magnitude is an ungodly amount of memory?

It seems to be noticeably more than 500MB (500MB is the amount of ram
on the computer that I use for the tests). It starts swapping and
instead of 1 hour for testing 2.5, 2.7 and 3.2 it takes ~3 hours.

I will check the new version now.

**Comment by Krastanov** [comment]
by the way, the problem may have been triggered by some failing test
(I have hash randomization on)

**Comment by asmeurer** [comment]
@Krastanov I think that's definitely the tests.  c.f. the [mailing list](https://groups.google.com/forum/?fromgroups#!topic/sympy/UP46rCtHZPg).

**Comment by asmeurer** [comment]
I need to test the new config stuff, but other than that, and my comment on the most recent commit, I think this is ready to go.

**Comment by asmeurer** [comment]
@flacjacket any preference as to which PR of yours gets merged first.  There is going to be a merge conflict either way.

**Comment by flacjacket** [comment]
I removed that last commit.

I think this one should be merged first. I think the configuration file writing should be re-worked with ConfigParser before it goes in.

File: README.rst [hunk 2]
```diff
@@ -46,26 +52,47 @@ connections.
...
+
+    interpreter = /path/to/python, /path/to/other/python
+
+If you pass the ``-3`` flag, the interpreters will be read from the
+``interpreter3`` option. Specifying ``interpreter = None`` you can disable the
+Python tests, which can be useful in setting up a profile just for testing
+docs.
+
+Any of the other options set by commandline parameters can be set in the
+configuration file. See ``sympy-bot --help`` for more information.
```
Comment (asmeurer) [inline README.rst]: This isn't very clear to me.  For example, what should I add to the config file to make a profile for testing the docs?

**Comment by asmeurer** [comment]
Is there a way to make one profile inherit and merge the settings from another profile, so that I don't have to duplicate the paths to all my Python executables many times?

**Comment by asmeurer** [comment]
Also, is it possible to add comments to the config file?

**Comment by asmeurer** [comment]
OK, looking at http://docs.python.org/library/configparser.html, I think I might be able to use the built-in stuff to do what I want.  Comments are delineated by `#` or `;`.

ConfigParser automatically does new-style string formatting type expansion, like

```
var = value
f = %(var)s
```

and `f` will be set to `value`.  What I want is an easy way to type all my Python executables once (I have about a dozen of them).  I'll see how easy it is, and post a link to my config file if I get it working.

**Comment by asmeurer** [comment]
Hmm.  It seems that nested replacements are not supported (or else I did it wrong).  This is what I've got: https://github.com/asmeurer/dotfiles/blob/master/.sympy/sympy-bot.conf.

File: README.rst [hunk 3]
```diff
@@ -29,6 +29,12 @@ to only review mergeable pull requests, do::
 
     ./sympy-bot review mergeable
 
+Requirements
+------------
+
+SymPy bot needs argparse to run. This is part of the standard library in
+Python 2.7 and 3.2, however it can be installed in earlier versions of Python.
```
Comment (asmeurer) [inline README.rst]: This suggests that sympy-bot supports Python 3

**Comment by asmeurer** [comment]
Oh, I see what happened.  I incorrectly assumed that all the values from `[default]` would be pulled in.  It seems to be impossible to have values not in a profile, so can this be fixed?

File: sympy-bot [hunk 4]
```diff
@@ -1,362 +1,471 @@
...
-    else:
-        print "Too many arguments"
-        sys.exit(1)
-    parser.print_help()
+                parser.readfp(f)
+            except IOError:
+                print "> WARNING: Unable to open config file"
+            except ConfigParser.Error:
+                print "> WARNING: Unable to parse config file"
+            finally:
```
Comment (asmeurer) [inline sympy-bot]: Should this rather be an else clause?

**Comment by asmeurer** [comment]
OK, I see why it's doing this. I'll push a fix to flacjacket/sympy-bot#1.

**Comment by asmeurer** [comment]
Another bug:

``` pytb
> View logs for PR 1427 in: /var/folders/wc/dppcpmxs1tlb36nqcw853wkm0000gq/T/sympy-bot-tmpWj2xcY/out
Traceback (most recent call last):
  File "./sympy-bot", line 473, in <module>
    main()
  File "./sympy-bot", line 175, in main
    dispatch_reviews(options, urls, username=username, password=password)
  File "./sympy-bot", line 375, in dispatch_reviews
    config.build_docs, config.build_docs_command, user)
  File "./sympy-bot", line 429, in formulate_review
    details = get_platform_version(i)
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy-bot/utils/cmd.py", line 97, in get_platform_version
    executable = get_executable(interpreter)
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy-bot/utils/cmd.py", line 80, in get_executable
    return ouput.strip()
NameError: global name 'ouput' is not defined
```

File: sympy-bot [hunk 5]
```diff
@@ -1,362 +1,471 @@
...
+                print "> Loaded configuration file"
+
+                if parser.has_section(profile):
+                    items = dict(parser.items(profile))
+                else:
+                    print "> WARNING: Configuration file does not contain profile %s" % profile
```
Comment (asmeurer) [inline sympy-bot]: IMHO, this should be a fatal error.

**Comment by asmeurer** [comment]
This is not doing the right thing at all for the `interpreter3` flag in the config file.  It ignores it for profiles that have both it and `interpreter`, and it uses the default interpreter for those that don't have any `interpreter`.

File: sympy-bot (hunk 4 above)
Comment (flacjacket) [inline sympy-bot]: Yes, definitely.

**Comment by flacjacket** [comment]
The `interpreter3` configuration option is working for me. The interpreters used are first taken from the `-i` options, and if none are specified, falls back to `interpreter3` config if `-3` is passed and `interpreter` config otherwise.

**Comment by asmeurer** [comment]
Is there a way to add -3 to the profile them?  Personally I think that's somewhat confusing.

**Comment by asmeurer** [comment]
Actually what I want to do is run both Python 2 and Python 3 tests in one go.  Is there a way to do that?

**Comment by flacjacket** [comment]
-3 is the same as --python3, so you can set it with `python3 = True`.

I think the best way to do that would be to have `-i` use `interpreter` (which is the default, and so only necessary if `-D` is passed), `-3` use `interpreter3`, and `-i3` use both `interpreter` and `interpreter3`. Currently `-i` and `-3` work like this, I'd have to add the `-i3` behavior if you think that is good.

**Comment by asmeurer** [comment]
I don't really care about command line options.  I just want a way to make a profile that runs both Python 2 and Python 3 tests.

**Comment by asmeurer** [comment]
I guess the problem is the dual role that the interpreter option plays.  Maybe we should add a -2/--python2 option, at least for the sake of the config file.

**Comment by flacjacket** [comment]
Yes, my thought exactly. I was just in the process of adding a -2 option.

**Comment by flacjacket** [comment]
I just pushed a commit that should do the trick, setting `python2 = True` and `python3 = True`.

**Comment by asmeurer** [comment]
OK, let me test it.  If it works, then I think this is ready to go.

**Comment by asmeurer** [comment]
A problem

``` pytb
Traceback (most recent call last):
  File "./sympy-bot", line 478, in <module>
    main()
  File "./sympy-bot", line 184, in main
    dispatch_reviews(options, urls, username=username, password=password)
  File "./sympy-bot", line 276, in dispatch_reviews
    python3[i], config.master_commit)
NameError: global name 'python3' is not defined
```

**Comment by flacjacket** [comment]
Gah, that's what I get for running out the door without being able to test. I amended that commit so it works now.

**Comment by asmeurer** [comment]
I guess this is unrelated to this branch, but we should remove all the duplicate work for multiple testing sessions, like cloning, use2to3, etc.

**Comment by flacjacket** [comment]
I think leaving the same clone should be fine. It will re-checkout the master branch and delete the test branch after each test run. This commit will delete the py3k-sympy directory after each Python 3 run.

**Comment by asmeurer** [comment]
Ah I see my previous comment was not worded the best. I meant we should remove the _duplication_ of work. In other words, don't clone, merge, use2to3, etc. more than once per run. use2to3 especially is a waste of time.

**Comment by asmeurer** [comment]
Once you get that cleaned up, this can be merged (after I test it of course).

**Comment by asmeurer** [comment]
There seems to be some inconstancies regarding having a profile that only runs Python 3 tests.  Apparently the correct way to do this is to set `interpreter = None`.  At first I thought that just having `interpreter3` set would do this, though I can see why it wouldn't.  But even more inconsistant is that setting `python2 = False` does not work either.

**Comment by asmeurer** [comment]
The `-i` option seems to be broken.  I did `sympy-bot review 1431 -i python2.5`, but it ran the tests with the default interpreter.

**Comment by asmeurer** [comment]
And it seems to me that we should have an `-i3` flag.  Otherwise you have to use the config file to run a non-default Python 3.

**Comment by asmeurer** [comment]
At https://github.com/sympy/sympy/pull/1431#issuecomment-7193726 you can see that several of the reports have `interpreter = None` in the summary.

**Comment by flacjacket** [comment]
Hm, do you know what the command was that gave that?

Also, it looks like some of the logic was wrong so it would never get a command line-specified interpreter. I think, however, the way it is currently has too many weird corner cases.

What I'd like to do is have `-2`/`python2 = True`, `-3`/`python3=True`, `-D`/`build_docs = True` and `-i` act separately. Where `-2`/`python2 = True` (`-3`/`python3 = True`) enables/disables the `interpreter`'s (`interpreter3`'s) in the config file, which will default to `python` (`python3`) (`-2` takes the place of just doing a bare `-i`). `-D`/`build_docs` builds the docs. Then `-i` can add additional interpreters, either Python 2 or 3 from the command line. If none of these 4 are specified, it does the default `interpreter` in the config file (falling back to `python`), and if any are specified, it only does those specified. This way they can all act independently, which is currently not the case.

Also, there is currently no checking whether the `interpreter3` interpreters are actually Python 3, is this a problem? Since it's just used for testing, I don't think it's necessary.

**Comment by asmeurer** [comment]
> Hm, do you know what the command was that gave that?

`./sympy-bot review 1431 --profile all-tests`, and my profile is at https://github.com/asmeurer/dotfiles/blob/master/.sympy/sympy-bot.conf (actually a little earlier in the history if that makes a difference).

> What I'd like to do is have -2/python2 = True, -3/python3=True, -D/build_docs = True and -i act separately. Where -2/python2 = True (-3/python3 = True) enables/disables the interpreter's (interpreter3's) in the config file, which will default to python (python3) (-2 takes the place of just doing a bare -i). -D/build_docs builds the docs. Then -i can add additional interpreters, either Python 2 or 3 from the command line. If none of these 4 are specified, it does the default interpreter in the config file (falling back to python), and if any are specified, it only does those specified. This way they can all act independently, which is currently not the case.

Well it's definitely confusing to me.  I think perhaps the confusing part is that `-i` acts like `-i python`.  Now that we have `-2`, maybe the bare `-i` should be removed.  

The way I see it, `-2`, `-3`, and `-D` are actions.  They specify to run certain tests.  `-i` and my proposed `-i3` on the other hand serve to list interpreters.  They are not actions.  `interpreter = None` runs no Python 2 tests not because it disables them, but because it sets the interpreter list to be run to the empty list.  

So the logic that makes sense to me is to have the following defaults

```
interpreter = python
interpreter3 = python3
python2 = True
python3 = False
build_docs = False
```

Changing any of these variables overrides the default.  To reiterate, `python2`, `python3`, and `build_docs` specify what tests will be run.  `interpreter` and `interpreter3` specify the interpreter(s) for `python2` and `python3`, respectively.  They can also be None, in which case those tests will not be run regardless of the values of `python2` and `python3` (or as we mathematicians like to say, they will be run, but on zero interpreters).

In order for this to work, we need a way to disable `python2`, and `python3`, perhaps by adding complementary `--no-python2` and `--no-python3` commands (these would not be mirrored in the config file, but would just be the way of setting the respective variables to `False`).  Or we could have `-2 False`, which would mirror the way things work in the config file, but might be a little more confusing.

And except for that point about `--no-python2/3`, all of this would be identical for command line options and the config file, where the former overrides the latter if there are any discrepancies. 

The question then is what to do about commands that automatically enable or disable other commands.  For example, `-D` automatically disables `-2` and `-3`.  Another one that we might want if we do things as I propose is to make `-3` disable `-2`, so that `sympy-bot review 123 -3` runs only the Python 3 tests and not the Python 2 tests.  These make things easier from the command line perspective, but more confusing, especially from the config file perspective.  I'm not sure what the best way here is.  Both ways are potentially confusing.  Thoughts? 

And for the rest as well, how does this sound to you?

> Also, there is currently no checking whether the interpreter3 interpreters are actually Python 3, is this a problem? Since it's just used for testing, I don't think it's necessary.

No.  I've noticed a dozen other type checks that don't happen either (e.g., I previously mistyped the path to one of my Pythons in my config file and it caused sympy-bot to die when it reached that particular interpreter). As long as you can tell upon running the bot that there is a misconfiguration, I think it's fine.

**Comment by flacjacket** [comment]
> And it seems to me that we should have an -i3 flag. Otherwise you have to use the config file to run a non-default Python 3.

Yeah, there should probably be an `--interpreter3` option. But it could only have a one character short flag, `-i3` would be read as `-i -3`.

> Now that we have -2, maybe the bare -i should be removed.

Yes, definitely.

> -i and my proposed -i3 on the other hand serve to list interpreters.

Would this be just printing the `interpreter` and `interpreter3` values and exiting?

I think having `--python2/3/build_docs=False` is more straightforward. We couldn't use the easy `store_true` argparse action, but the extra logic is probably on par with having 2 different flags and setting something like `python2 = False` will actually work. Plus then you know exactly what is being set by the command line.

I agree with having `-2/-3/-D` just enable/disable the actions. That's basically what I was trying to get at, but more clear. I think either `-3` or `-D` should disable the default enabling of `-2`. I feel making the configuration file a bit more verbose is alright, since it's created once, as opposed to complicating command line invocations.

**Comment by asmeurer** [comment]
> Would this be just printing the interpreter and interpreter3 values and exiting?

No.  I guess list was not the best word.  I just mean that these options are not actions.  Their only purpose is for the user to let sympy-bot know what interpreters to use when -2 or -3 are used.

**Comment by flacjacket** [comment]
Okay, this should now be fully functional as discussed. Let me know if you have any problems with it.

**Comment by asmeurer** [comment]
OK, I think it's working.  Anyway, I was able to run just the Python 3 tests from the command line, so that's good.

**Comment by asmeurer** [comment]
This is indeed working.  Thanks for all the hard work.  I'm merging…

Let me know when the other PR is ready to review.

**Comment by asmeurer** [comment]
I'm going to write to the mailing list about this.

**Pull request merged**